src/cpp/server/auth/ServerAuthHandler.cpp

```diff
@@ -358,25 +412,397 @@ bool isCookieRevoked(const std::string& cookie)
    return false;
 }
 
-namespace overlay {
-
-Error initialize()
+Error getUserFromDatabase(const boost::shared_ptr<IConnection>& connection,
+                          const system::User& user,
+                          bool* pLocked,
+                          boost::posix_time::ptime* pLastSignin,
+                          bool* pExists)
 {
+   *pLocked = true;
+
+   Rowset rows;
+   Query userQuery = connection->query("SELECT user_name, user_id, last_sign_in, locked FROM licensed_users WHERE user_id = :uid OR user_name = :username")
+         .withInput(user.getUserId())
+         .withInput(user.getUsername());
+
+   Error error = connection->execute(userQuery, rows);
+   if (error)
+      return error;
+
+   // Old versions of RSW didn't set the posix User ID. If we don't have the ID, update it  now.
+   bool foundUser = false;
+   for (RowsetIterator itr = rows.begin(); itr != rows.end(); ++itr)
+   {
+      Row& row = *itr;
+
+      int uid = row.get<int>(1);
+      const std::string& username = row.get<std::string>(0);
+      bool locked = row.get<int>(3) == 1;
+
+      // Check the last signin time. If we haven't seen a row for this user yet, update the time. 
+      // Otherwise, if there are somehow multiple rows for this user, update
+      // Ignore rows without a perfect username/uid match (except with uinitialized uids [<0])
+      // Safe static cast, because we already handled uid < 0
+      if ((username == user.getUsername()) && ((uid < 0) || (static_cast<UidType>(uid) == user.getUserId())))
+      {
+         boost::posix_time::ptime lastSignin = parseSigninTime(row.get<std::string>(2));
+         if (!foundUser || (lastSignin > *pLastSignin))
+            *pLastSignin = lastSignin;
+
+         // If any matching user row is not locked, the user is not locked.
+         *pLocked = (overlay::isUserLocked(locked) && *pLocked);
+
+         foundUser = true;
+      }
+
+      // If the user ID was uninitialized, update it now.
+      if (row.get<int>(1) < 0)
+      {
+         Query setUidQuery = connection->query("UPDATE licensed_users SET user_id = :uid WHERE user_name = :username and user_id = :olduid")
+            .withInput(user.getUserId())
+            .withInput(username)
+            .withInput(uid);
+
+         error = connection->execute(setUidQuery);
+         if (error)
+            return error;
+      }
+   }
+
+   *pExists = foundUser;
    return Success();
 }
 
-bool canStaySignedIn()
+bool isUserActive(const boost::posix_time::ptime& lastSignin)
 {
-   return true;
+   return (boost::posix_time::microsec_clock::universal_time() - lastSignin <
+           boost::posix_time::time_duration(kHoursInOneYear, 0, 0));
+}
+
+Error addUser(boost::asio::io_service& ioService,
+              const std::string& username,
+              bool isAdmin)
+{
+   boost::shared_ptr<IConnection> connection;
```

Comment:
If we are using the file system storage, should we maybe continue to not use the database?  I'm concerned that for OS users they are going to need to start administering the sqlite db, or at least needing to make sure it's operating properly even though they are getting no benefit. It might cause us less churn if this feature was tied to the DB for session metadata.

Reply:
There's already a requirement in OS that users have access to the database for revoked cookies, so I don't think that's a concern